File: runtime/cuda/CudaJitOptions.cpp
/*******************************************************************************
 * Copyright (c) 2013, 2015 IBM Corp. and others
 *
 * This program and the accompanying materials are made available under
 * the terms of the Eclipse Public License 2.0 which accompanies this
 * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
 * or the Apache License, Version 2.0 which accompanies this distribution and
 * is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * This Source Code may also be made available under the following
 * Secondary Licenses when the conditions for such availability set
 * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
 * General Public License, version 2 with the GNU Classpath
 * Exception [1] and GNU General Public License, version 2 with the
 * OpenJDK Assembly Exception [2].
 *
 * [1] https://www.gnu.org/software/classpath/license.html
 * [2] http://openjdk.java.net/legal/assembly-exception.html
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 *******************************************************************************/

#include "CudaCommon.hpp"
#include "java/com_ibm_cuda_CudaJitOptions.h"

namespace
{

/**
 * Destroy a JIT options instance.
 */
void
destroyOptions(JNIEnv * env, J9CudaJitOptions * options)
{
	J9VMThread * thread = (J9VMThread *)env;

	Trc_cuda_jitOptionsDestroy0_entry(thread, options);

	PORT_ACCESS_FROM_ENV(env);

	void * errorLog = options->errorLogBuffer;

	if (NULL != errorLog) {
		J9CUDA_FREE_MEMORY(errorLog);
	}

	void * infoLog = options->infoLogBuffer;

	if (NULL != infoLog) {
		J9CUDA_FREE_MEMORY(infoLog);
	}

	J9CUDA_FREE_MEMORY(options);

	Trc_cuda_jitOptionsDestroy0_exit(thread);
}

} // namespace

/**
 * Create and initialize a JIT options object.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    create
 * Signature: ([I)J
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] pairs     an array of option/value pairs
 * @return a pointer to a new JIT options object
 */
jlong JNICALL
Java_com_ibm_cuda_CudaJitOptions_create
  (JNIEnv * env, jclass, jintArray pairs)
{
	J9VMThread * thread = (J9VMThread *)env;

	Trc_cuda_jitOptionsCreate_entry(thread, pairs);

	jsize const   pairLength = env->GetArrayLength(pairs) & ~1;
	jint        * pairData   = NULL;

	if (0 != pairLength) {
		pairData = env->GetIntArrayElements(pairs, NULL);

		if (NULL == pairData) {
			Trc_cuda_jitOptionsCreate_getFail(thread);
			throwCudaException(env, J9CUDA_ERROR_OPERATING_SYSTEM);
			Trc_cuda_jitOptionsCreate_exit(thread, NULL);
			return 0;
		}
	}

	PORT_ACCESS_FROM_ENV(env);

	int32_t            error   = 0;
	J9CudaJitOptions * options = (J9CudaJitOptions *)J9CUDA_ALLOCATE_MEMORY(sizeof(J9CudaJitOptions));

	if (NULL == options) {
		Trc_cuda_jitOptionsCreate_allocFail(thread);
		error = J9CUDA_ERROR_MEMORY_ALLOCATION;
	} else {
		memset(options, 0, sizeof(J9CudaJitOptions));

		for (jint index = 0; (0 == error) && (index < pairLength); index += 2) {
			jint key   = pairData[index];
			jint value = pairData[index + 1];

			switch (key) {
			case com_ibm_cuda_CudaJitOptions_OPT_CACHE_MODE:
				options->cacheMode = (J9CudaJitCacheMode)value;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_ERROR_LOG_BUFFER_SIZE_BYTES:
				if (NULL != options->errorLogBuffer) {
					J9CUDA_FREE_MEMORY(options->errorLogBuffer);
				}

				options->errorLogBuffer = (char *)J9CUDA_ALLOCATE_MEMORY((size_t)value);
				options->errorLogBufferSize = (uintptr_t)value;

				if (NULL == options->errorLogBuffer) {
					Trc_cuda_jitOptionsCreate_allocFail(thread);
					options->errorLogBufferSize = 0;
					error = J9CUDA_ERROR_MEMORY_ALLOCATION;
				}
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_FALLBACK_STRATEGY:
				options->fallbackStrategy = (J9CudaJitFallbackStrategy)value;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_GENERATE_DEBUG_INFO:
				options->generateDebugInfo = value ? J9CUDA_JIT_FLAG_ENABLED : J9CUDA_JIT_FLAG_DISABLED;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_GENERATE_LINE_INFO:
				options->generateLineInfo = value ? J9CUDA_JIT_FLAG_ENABLED : J9CUDA_JIT_FLAG_DISABLED;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_INFO_LOG_BUFFER_SIZE_BYTES:
				if (NULL != options->infoLogBuffer) {
					J9CUDA_FREE_MEMORY(options->infoLogBuffer);
				}

				options->infoLogBuffer = (char *)J9CUDA_ALLOCATE_MEMORY((size_t)value);
				options->infoLogBufferSize = (uintptr_t)value;

				if (NULL == options->infoLogBuffer) {
					Trc_cuda_jitOptionsCreate_allocFail(thread);
					options->infoLogBufferSize = 0;
					error = J9CUDA_ERROR_MEMORY_ALLOCATION;
				}
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_LOG_VERBOSE:
				options->verboseLogging = value ? J9CUDA_JIT_FLAG_ENABLED : J9CUDA_JIT_FLAG_DISABLED;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_MAX_REGISTERS:
				options->maxRegisters = (uint32_t)value;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_OPTIMIZATION_LEVEL:
				switch (value) {
				case 0:
					options->optimizationLevel = J9CUDA_JIT_OPTIMIZATION_LEVEL_0;
					break;
				case 1:
					options->optimizationLevel = J9CUDA_JIT_OPTIMIZATION_LEVEL_1;
					break;
				case 2:
					options->optimizationLevel = J9CUDA_JIT_OPTIMIZATION_LEVEL_2;
					break;
				case 3:
					options->optimizationLevel = J9CUDA_JIT_OPTIMIZATION_LEVEL_3;
					break;
				case 4:
					options->optimizationLevel = J9CUDA_JIT_OPTIMIZATION_LEVEL_4;
					break;
				default:
					error = J9CUDA_ERROR_INVALID_VALUE;
					break;
				}
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_TARGET:
				options->target = (uint32_t)value;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_TARGET_FROM_CUCONTEXT:
				options->targetFromContext = value ? J9CUDA_JIT_FLAG_ENABLED : J9CUDA_JIT_FLAG_DISABLED;
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_THREADS_PER_BLOCK:
				if (0 != value) {
					options->threadsPerBlock = (uint32_t)value;
				} else {
					error = J9CUDA_ERROR_INVALID_VALUE;
				}
				break;

			case com_ibm_cuda_CudaJitOptions_OPT_WALL_TIME:
				options->recordWallTime = value ? J9CUDA_JIT_FLAG_ENABLED : J9CUDA_JIT_FLAG_DISABLED;
				break;

			default:
				Trc_cuda_jitOptionsCreate_badOption(thread, key);
				error = J9CUDA_ERROR_INVALID_VALUE;
				break;
			}
		}
	}

	if (NULL != pairData) {
		env->ReleaseIntArrayElements(pairs, pairData, JNI_ABORT);
	}

	if (0 != error) {
		throwCudaException(env, error);
	}

	if (env->ExceptionCheck() && (NULL != options)) {
		destroyOptions(env, options);
		options = NULL;
	}

	Trc_cuda_jitOptionsCreate_exit(thread, options);

	return (jlong)options;
}

/**
 * Destroy a JIT options object.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    destroy
 * Signature: (J)V
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] handle    the options pointer
 */
void JNICALL
Java_com_ibm_cuda_CudaJitOptions_destroy
  (JNIEnv * env, jclass, jlong handle)
{
	J9VMThread       * thread  = (J9VMThread *)env;
	J9CudaJitOptions * options = (J9CudaJitOptions *)handle;

	Trc_cuda_jitOptionsDestroy_entry(thread, options);

	destroyOptions(env, options);

	Trc_cuda_jitOptionsDestroy_exit(thread);
}

/**
 * Return the contents of the error log.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    getErrorLogBuffer
 * Signature: (J)Ljava/lang/String;
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] handle    the options pointer
 * @return a string containing the error log
 */
jstring JNICALL
Java_com_ibm_cuda_CudaJitOptions_getErrorLogBuffer
  (JNIEnv * env, jclass, jlong handle)
{
	J9VMThread       * thread  = (J9VMThread *)env;
	J9CudaJitOptions * options = (J9CudaJitOptions *)handle;

	Trc_cuda_jitOptionsGetErrorLogBuffer_entry(thread, options);

	const char * buffer = ((J9CudaJitOptions *)handle)->errorLogBuffer;
	jstring      result = env->NewStringUTF((NULL == buffer) ? "" : buffer);

	Trc_cuda_jitOptionsGetErrorLogBuffer_exit(thread, result);

	return result;
}

/**
 * Return the contents of the information log.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    getInfoLogBuffer
 * Signature: (J)Ljava/lang/String;
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] handle    the options pointer
 * @return a string containing the information log
 */
jstring JNICALL
Java_com_ibm_cuda_CudaJitOptions_getInfoLogBuffer
  (JNIEnv * env, jclass, jlong handle)
{
	J9VMThread         * thread   = (J9VMThread *)env;
	J9CudaJitOptions * options = (J9CudaJitOptions *)handle;

	Trc_cuda_jitOptionsGetInfoLogBuffer_entry(thread, options);

	const char * buffer = options->infoLogBuffer;
	jstring      result = env->NewStringUTF((NULL == buffer) ? "" : buffer);

	Trc_cuda_jitOptionsGetInfoLogBuffer_exit(thread, result);

	return result;
}

/**
 * Return the number of threads the compiler actually targeted per block.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    getThreadsPerBlock
 * Signature: (J)I
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] handle    the options pointer
 * @return the number of threads per block
 */
jint JNICALL
Java_com_ibm_cuda_CudaJitOptions_getThreadsPerBlock
  (JNIEnv * env, jclass, jlong handle)
{
	J9VMThread         * thread   = (J9VMThread *)env;
	J9CudaJitOptions * options = (J9CudaJitOptions *)handle;

	Trc_cuda_jitOptionsGetThreadsPerBlock_entry(thread, options);

	jint result = (jint)options->threadsPerBlock;

	Trc_cuda_jitOptionsGetThreadsPerBlock_exit(thread, result);

	return result;
}

/**
 * Return the elapsed compile or link time in milliseconds.
 *
 * Class:     com.ibm.cuda.CudaJitOptions
 * Method:    getWallTime
 * Signature: (J)F
 *
 * @param[in] env       the JNI interface pointer
 * @param[in] (unused)  the class pointer
 * @param[in] handle    the options pointer
 * @return the elapsed compile/link time
 */
jfloat JNICALL
Java_com_ibm_cuda_CudaJitOptions_getWallTime
  (JNIEnv * env, jclass, jlong handle)
{
	J9VMThread         * thread   = (J9VMThread *)env;
	J9CudaJitOptions * options = (J9CudaJitOptions *)handle;

	Trc_cuda_jitOptionsGetWallTime_entry(thread, options);

	jfloat result = (jfloat)options->wallTime;

	Trc_cuda_jitOptionsGetWallTime_exit(thread, result);

	return result;
}

File: jcl/src/openj9.cuda/share/classes/com/ibm/cuda/CudaJitOptions.java
/*[INCLUDE-IF Sidecar18-SE]*/
/*******************************************************************************
 * Copyright (c) 2013, 2018 IBM Corp. and others
 *
 * This program and the accompanying materials are made available under
 * the terms of the Eclipse Public License 2.0 which accompanies this
 * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
 * or the Apache License, Version 2.0 which accompanies this distribution and
 * is available at https://www.apache.org/licenses/LICENSE-2.0.
 *
 * This Source Code may also be made available under the following
 * Secondary Licenses when the conditions for such availability set
 * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
 * General Public License, version 2 with the GNU Classpath
 * Exception [1] and GNU General Public License, version 2 with the
 * OpenJDK Assembly Exception [2].
 *
 * [1] https://www.gnu.org/software/classpath/license.html
 * [2] http://openjdk.java.net/legal/assembly-exception.html
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
 *******************************************************************************/
package com.ibm.cuda;

import java.util.concurrent.atomic.AtomicLong;

/**
 * The {@code CudaJitOptions} class represents a set of options that influence
 * the behavior of linking and loading modules.
 */
public final class CudaJitOptions implements Cloneable {

	/**
	 * {@code CacheMode} identifies the cache management choices.
	 */
	public static enum CacheMode {

		/** Compile with no -dlcm flag specified. */
		DEFAULT(1),

		/** Compile with L1 cache disabled. */
		L1_DISABLED(2),

		/** Compile with L1 cache enabled. */
		L1_ENABLED(3);

		final int nativeMode;

		CacheMode(int nativeMode) {
			this.nativeMode = nativeMode;
		}
	}

	/**
	 * {@code Fallback} identifies the available fall-back strategies
	 * when an exactly matching object is not available.
	 */
	public static enum Fallback {

		/** Prefer to fall back to compatible binary code if exact match not found. */
		PreferBinary(1),

		/** Prefer to compile ptx if exact binary match not found. */
		PreferPtx(2);

		final int nativeStrategy;

		Fallback(int nativeStrategy) {
			this.nativeStrategy = nativeStrategy;
		}
	}

	// The range of OPT_* constants.
	private static final int NUM_OPT_CODES = 15;

	private static final int OPT_CACHE_MODE = 14;

	private static final int OPT_ERROR_LOG_BUFFER_SIZE_BYTES = 6;

	private static final int OPT_FALLBACK_STRATEGY = 10;

	private static final int OPT_GENERATE_DEBUG_INFO = 11;

	private static final int OPT_GENERATE_LINE_INFO = 13;

	private static final int OPT_INFO_LOG_BUFFER_SIZE_BYTES = 4;

	private static final int OPT_LOG_VERBOSE = 12;

	private static final int OPT_MAX_REGISTERS = 0;

	private static final int OPT_OPTIMIZATION_LEVEL = 7;

	private static final int OPT_TARGET = 9;

	private static final int OPT_TARGET_FROM_CUCONTEXT = 8;

	private static final int OPT_THREADS_PER_BLOCK = 1;

	private static final int OPT_WALL_TIME = 2;

	private static native long create(int[] keyValuePairs) throws CudaException;

	private static native void destroy(long handle);

	private static native String getErrorLogBuffer(long handle);

	private static native String getInfoLogBuffer(long handle);

	private static native int getThreadsPerBlock(long handle);

	private static native float getWallTime(long handle);

	private String errorLogBuffer;

	private String infoLogBuffer;

	private final AtomicLong nativeHandle;

	private int optionMask;

	private final int[] optionValue;

	private int threadsPerBlock;

	private float wallTime;

	/**
	 * Creates a new options object.
	 */
	public CudaJitOptions() {
		super();
		this.errorLogBuffer = ""; //$NON-NLS-1$
		this.infoLogBuffer = ""; //$NON-NLS-1$
		this.nativeHandle = new AtomicLong();
		this.optionMask = 0;
		this.optionValue = new int[NUM_OPT_CODES];
		this.threadsPerBlock = 0;
		this.wallTime = 0.0f;
	}

	/**
	 * Creates a new options object with the same state as this object.
	 */
	@Override
	protected CudaJitOptions clone() {
		CudaJitOptions clone = new CudaJitOptions();

		clone.optionMask = optionMask;
		System.arraycopy(optionValue, 0, clone.optionValue, 0, NUM_OPT_CODES);

		return clone;
	}

	/**
	 * Returns the contents of the error log.
	 * <p>
	 * The result will be empty unless {@link #setErrorLogBufferSize(int)}
	 * was called with a positive value, this object was used in connection
	 * with a {@link CudaModule} or a {@link CudaLinker}, and errors were
	 * reported.
	 *
	 * @return
	 *          the contents of the error log
	 */
	public String getErrorLogBuffer() {
		return errorLogBuffer;
	}

	long getHandle() throws CudaException {
		long handle;

		while ((handle = nativeHandle.get()) == 0) {
			int mask = optionMask;
			int[] keyValuePairs = new int[Integer.bitCount(mask) << 1];
			int index = 0;

			for (int code = 0; mask != 0; ++code, mask >>= 1) {
				if ((mask & 1) != 0) {
					keyValuePairs[index++] = code;
					keyValuePairs[index++] = optionValue[code];
				}
			}

			handle = create(keyValuePairs);

			if (nativeHandle.compareAndSet(0, handle)) {
				break;
			}

			// forget our work and look for results from a thread that finished earlier
			destroy(handle);
		}

		return handle;
	}

	/**
	 * Returns the contents of the information log.
	 * <p>
	 * The result will be empty unless {@link #setInfoLogBufferSize(int)}
	 * was called with a positive value, this object was used in connection
	 * with a {@link CudaModule} or a {@link CudaLinker}, and informational
	 * messages were reported.
	 *
	 * @return
	 *          the contents of the information log
	 */
	public String getInfoLogBuffer() {
		return infoLogBuffer;
	}

	/**
	 * Returns the maximum number of threads per block.
	 * <p>
	 * The result will only be meaningful if {@link #setThreadsPerBlock(int)} was
	 * called with a positive value, and this object was used in connection
	 * with a {@link CudaModule} or a {@link CudaLinker} involving PTX code.
	 *
	 * @return
	 *          the maximum number of threads per block
	 */
	public int getThreadsPerBlock() {
		return threadsPerBlock;
	}

	/**
	 * Returns the total elapsed time, in milliseconds,
	 * spent in the compiler and linker.
	 * <p>
	 * Applies to: compiler and linker.
	 *
	 * @return
	 *          the total elapsed time, in milliseconds, spent in the compiler and linker
	 */
	public float getWallTime() {
		return wallTime;
	}

	/**
	 * Requests recording of the total wall clock time,
	 * in milliseconds, spent in the compiler and linker.
	 * <p>
	 * Applies to: compiler and linker.
	 *
	 * @return
	 *          this options object
	 */
	public CudaJitOptions recordWallTime() {
		return setOption(OPT_WALL_TIME, true);
	}

	void releaseHandle(boolean update) {
		long handle = nativeHandle.getAndSet(0);

		if (handle != 0) {
			if (update) {
				update(handle);
			}

			destroy(handle);
		}
	}

	/**
	 * Specifies the desired caching behavior (-dlcm).
	 * <p>
	 * Applies to compiler only.
	 *
	 * @param mode
	 *          the desired caching behavior
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setCacheMode(CacheMode mode) {
		if (mode == null) {
			mode = CacheMode.DEFAULT;
		}

		return setOption(OPT_CACHE_MODE, mode.nativeMode);
	}

	/**
	 * Specifies the size, in bytes, to allocate for capturing error messages.
	 * <p>
	 * Applies to compiler and linker.
	 *
	 * @param size
	 *          the size, in bytes, of the error log buffer
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setErrorLogBufferSize(int size) {
		return setOption(OPT_ERROR_LOG_BUFFER_SIZE_BYTES, size);
	}

	/**
	 * Specifies whether to generate debug information.
	 * <p>
	 * Applies to compiler and linker.
	 *
	 * @param enabled
	 *          whether debug information should be generated
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setGenerateDebugInfo(boolean enabled) {
		return setOption(OPT_GENERATE_DEBUG_INFO, enabled);
	}

	/**
	 * Specifies whether to generate line number information.
	 * <p>
	 * Applies to compiler only.
	 *
	 * @param enabled
	 *          whether line number information should be generated
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setGenerateLineInfo(boolean enabled) {
		return setOption(OPT_GENERATE_LINE_INFO, enabled);
	}

	/**
	 * Specifies the size, in bytes, to allocate for capturing informational
	 * messages.
	 * <p>
	 * Applies to compiler and linker.
	 *
	 * @param size
	 *          the size, in bytes, of the information log buffer
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setInfoLogBufferSize(int size) {
		return setOption(OPT_INFO_LOG_BUFFER_SIZE_BYTES, size);
	}

	/**
	 * Specifies the fallback strategy if an exactly matching
	 * binary object cannot be found.
	 * <p>
	 * Applies to: compiler only
	 *
	 * @param strategy
	 *          the desired fallback strategy
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setJitFallbackStrategy(Fallback strategy) {
		if (strategy == null) {
			strategy = Fallback.PreferPtx;
		}

		return setOption(OPT_FALLBACK_STRATEGY, strategy.nativeStrategy);
	}

	/**
	 * Specifies whether to generate verbose log messages.
	 * <p>
	 * Applies to: compiler and linker
	 *
	 * @param verbose
	 *          whether verbose log messages should be generated
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setLogVerbose(boolean verbose) {
		return setOption(OPT_LOG_VERBOSE, verbose);
	}

	/**
	 * Specifies the maximum number of registers that a thread may use.
	 * <p>
	 * Applies to: compiler only
	 *
	 * @param limit
	 *          the maximum number of registers a thread may use
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setMaxRegisters(int limit) {
		return setOption(OPT_MAX_REGISTERS, limit);
	}

	/**
	 * Specifies the level of optimization to be applied to generated code
	 * (0 - 4), with 4 being the default and highest level of optimization.
	 * <p>
	 * Applies to compiler only.
	 *
	 * @param level
	 *          the desired optimization level
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setOptimizationLevel(int level) {
		return setOption(OPT_OPTIMIZATION_LEVEL, level);
	}

	private CudaJitOptions setOption(int key, boolean value) {
		return setOption(key, value ? 1 : 0);
	}

	private CudaJitOptions setOption(int key, int value) {
		releaseHandle(false);

		optionMask |= 1 << key;
		optionValue[key] = value;

		return this;
	}

	/**
	 * Specifies the desired compute target.
	 * <p>
	 * Cannot be combined with {@link #setThreadsPerBlock(int)}.
	 * <p>
	 * Applies to compiler and linker.
	 *
	 * @param target
	 *          the desired compute target
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setTarget(CudaJitTarget target) {
		return setOption(OPT_TARGET, target.nativeValue);
	}

	/**
	 * Specifies that the target should be determined based on the current
	 * attached context.
	 * <p>
	 * Applies to compiler and linker.
	 *
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setTargetFromCuContext() {
		return setOption(OPT_TARGET_FROM_CUCONTEXT, true);
	}

	/**
	 * Specifies the minimum number of threads per block for compilation.
	 * <p>
	 * This restricts the resource utilization of the compiler (e.g. maximum
	 * registers) such that a block with the given number of threads should be
	 * able to launch based on register limitations. Note, this option does not
	 * currently take into account any other resource limitations, such as
	 * shared memory utilization.
	 * <p>
	 * Cannot be combined with {@link #setTarget(CudaJitTarget)}.
	 * <p>
	 * Applies to compiler only.
	 *
	 * @param limit
	 *          the desired minimum number of threads per block
	 * @return
	 *          this options object
	 */
	public CudaJitOptions setThreadsPerBlock(int limit) {
		return setOption(OPT_THREADS_PER_BLOCK, limit);
	}

	CudaJitOptions update() {
		long handle = nativeHandle.get();

		if (handle != 0) {
			update(handle);
		}

		return this;
	}

	private void update(long handle) {
		errorLogBuffer = getErrorLogBuffer(handle);
		infoLogBuffer = getInfoLogBuffer(handle);
		threadsPerBlock = getThreadsPerBlock(handle);
		wallTime = getWallTime(handle);
	}
}
